zkvm/src/signature/multikey.rs

```diff
@@ -0,0 +1,64 @@
+use crate::errors::VMError;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+
+#[derive(Clone)]
+pub struct Multikey {
+    transcript: Transcript,
+    X_agg: VerificationKey,
+    L: Scalar,
+}
+
+impl Multikey {
+    pub fn new(pubkeys: Vec<VerificationKey>) -> Result<Self, VMError> {
+        // Create transcript for Multikey
+        let mut transcript = Transcript::new(b"ZkVM.aggregated-key");
+
+        // Hash in pubkeys
+        // L = H(X_1 || X_2 || ... || X_n)
+        for X_i in &pubkeys {
+            transcript.commit_point(b"X_i.L", &X_i.0);
+        }
+        let L = transcript.challenge_scalar(b"L");
+
+        // Make aggregated pubkey
+        // X = sum_i ( a_i * X_i )
+        // a_i = H(L, X_i). Compenents of L have already been fed to transcript.
+        let mut X_agg = RistrettoPoint::default();
+        for X_i in &pubkeys {
+            let mut a_i_transcript = transcript.clone();
+            a_i_transcript.commit_point(b"X_i", &X_i.0);
+            let a_i = a_i_transcript.challenge_scalar(b"a_i");
```

Comment:
to avoid duplication we can use `self.a_i()` method (aka `self.factor_for_key`).

Reply:
We don't have `self` yet, since this is happening within the constructor function.
So, should we make a "temporary, half-formed" `multikey` and then call `self.factor_for_key()` on that? It seems messy, but I'll try it out... let me know what you think?

Reply:
Oleg suggested a solution here,
https://github.com/interstellar/slingshot/pull/190#discussion_r265362498
and I implemented it! resolved!